modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java

```diff
@@ -530,8 +531,22 @@ public SchedulerJobConfiguration addingService(
 			SchedulerJobConfiguration schedulerJobConfiguration =
 				_bundleContext.getService(serviceReference);
 
-			TriggerConfiguration triggerConfiguration =
-				schedulerJobConfiguration.getTriggerConfiguration();
+			TriggerConfiguration triggerConfiguration = null;
```

Comment:
@shuyangzhou the problem with constructing the TriggerConfiguration during creation/activation time is that the we can do that on our own JobConfiguration classes, but we cannot enforce this on the custom developments without creating a base class implying a breaking change.

So somehow I only see to enforce customer own jobs using a valid configuration:

- Leaving the check in the SchedulerEngineHelperImpl
- Creating a Base class to implement and create the TriggerConfiguration in the constructor. But this will imply a new breaking change.

Reply:
@shuyangzhou were you able to take a look to this concern?

Reply:
No need to enforce it in any way. The fact the getter declares no checked exception means we don't foresee it will happen. For runtime exceptions they can happen anywhere for any reasons, we can not cover them all.

Reply:
@shuyangzhou I still don't understand this point:

> so that a badly configured SchedulerJobConfiguration won't even show up as a service, rather we catch and log for a bad service later.

If I construct the TriggerConfiguration during the activate the service is going to show up as a service, but with state _FAILED ACTIVATION_

```
scr:info com.liferay.announcements.web.internal.scheduler.CheckEntrySchedulerJobConfiguration
Component Description: com.liferay.announcements.web.internal.scheduler.CheckEntrySchedulerJobConfiguration
===========================================================================================================
Class:         com.liferay.announcements.web.internal.scheduler.CheckEntrySchedulerJobConfiguration
Bundle:        1247 (com.liferay.announcements.web:6.0.53)
Enabled:       true
Immediate:     false
Services:      [com.liferay.portal.kernel.scheduler.SchedulerJobConfiguration]
Scope:         singleton
Config PID(s): [com.liferay.announcements.web.internal.scheduler.CheckEntrySchedulerJobConfiguration], Policy: optional
Base Props:    (0 entries)

Component Configuration Id: 10803
---------------------------------
State:        FAILED ACTIVATION
Service:      5221 [com.liferay.portal.kernel.scheduler.SchedulerJobConfiguration]
```

But still appear as a registered service:

```
b 1247
com.liferay.announcements.web_6.0.53 [1247]
  Id=1247, Status=ACTIVE      Data Root=/media/malvaro/datos1/dev/node1/osgi/state/org.eclipse.osgi/1247/data
  "Registered Services"
    ...
    {com.liferay.portal.kernel.scheduler.SchedulerJobConfiguration}={component.id=10803, component.name=com.liferay.announcements.web.internal.scheduler.CheckEntrySchedulerJobConfiguration, service.id=5221, service.scope=bundle, service.bundleid=1247}
```



And because the activation is triggered by the SchedulerHelperEngineImpl:
```

java.lang.IllegalArgumentException: Interval is either equal or less than 0
	at com.liferay.portal.kernel.scheduler.TriggerConfiguration.<init>(TriggerConfiguration.java:51) ~[portal-kernel.jar:?]
	at com.liferay.portal.kernel.scheduler.TriggerConfiguration.createTriggerConfiguration(TriggerConfiguration.java:20) ~[portal-kernel.jar:?]
	at com.liferay.announcements.web.internal.scheduler.CheckEntrySchedulerJobConfiguration.activate(CheckEntrySchedulerJobConfiguration.java:58) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:228) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:664) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:510) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:310) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:300) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:341) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:114) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:983) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:956) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:901) ~[?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse$1.run(ServiceFactoryUse.java:212) ~[org.eclipse.osgi.jar:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_131]
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:210) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:111) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:45) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:519) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:460) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:624) ~[org.eclipse.osgi.jar:?]
	at com.liferay.portal.scheduler.internal.SchedulerEngineHelperImpl$SchedulerJobConfigurationServiceTrackerCustomizer.addingService(SchedulerEngineHelperImpl.java:531) ~[?:?]
```

I will still need to perform a check in the ServiceTrackerCustomizer (check null) to avoid a RuntimeException that affects the whole scheduler.

Reply:
The main point is to not do the try-catch around ```BundleContext.getService()``` . When a service is registered, it is like a promise from provider that it is ready to be used. It is not the consumer's job to prevent a badly supplied service. If we do that kind of protection here, what about the other couple hundred explicit usages? (search by ```git grep "undleContext.getService("``` and there are also tons of other hidden usages from lib.

When you make change, you can not just only think for this case as if it is totally isolated just by itself. A bad example could spread by other people's copy/paste.

Reply:
@shuyangzhou I was not talking about surrounding `BundleContext.getService()` by a try-catch, I'm talking about checking for null, like this:

```
SchedulerJobConfiguration schedulerJobConfiguration =
    _bundleContext.getService(serviceReference);
			
if (schedulerJobConfiguration == null) {
    return null;
}

TriggerConfiguration triggerConfiguration =
    schedulerJobConfiguration.getTriggerConfiguration();
```

To avoid the NullPointerException:

```
java.lang.NullPointerException: null
	at com.liferay.portal.scheduler.internal.SchedulerEngineHelperImpl$SchedulerJobConfigurationServiceTrackerCustomizer.addingService(SchedulerEngineHelperImpl.java:534) ~[?:?]
	at com.liferay.portal.scheduler.internal.SchedulerEngineHelperImpl$SchedulerJobConfigurationServiceTrackerCustomizer.addingService(SchedulerEngineHelperImpl.java:522) ~[?:?]
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:321) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:264) ~[org.eclipse.osgi.jar:?]
	at com.liferay.osgi.util.ServiceTrackerFactory.open(ServiceTrackerFactory.java:95) ~[?:?]
	at com.liferay.portal.scheduler.internal.SchedulerEngineHelperImpl.lambda$activate$0(SchedulerEngineHelperImpl.java:347) ~[?:?]
	at com.liferay.portal.kernel.dependency.manager.DependencyManagerSyncUtil.lambda$registerSyncCallable$0(DependencyManagerSyncUtil.java:34) ~[portal-kernel.jar:?]
	at com.liferay.portal.kernel.dependency.manager.DependencyManagerSyncUtil$1.complete(DependencyManagerSyncUtil.java:93) [portal-kernel.jar:?]
	at com.liferay.portal.kernel.concurrent.DefaultNoticeableFuture$OnceFutureListener.complete(DefaultNoticeableFuture.java:102) [portal-kernel.jar:?]
```

However, even if I check for null, the following exception is obtained, besides the expected one (`java.lang.IllegalArgumentException: Interval is either equal or less than 0`):

```
2023-10-02 06:17:37.061 WARN  [Framework Event Dispatcher: Equinox Container: 2b6f412f-e05f-4f1c-b169-b4699d89b017][Framework:53] FrameworkEvent WARNING
org.osgi.framework.ServiceException: org.apache.felix.scr.impl.manager.SingleComponentManager.getService() returned a null service object
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:232) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:111) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:45) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:519) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:460) ~[org.eclipse.osgi.jar:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:624) ~[org.eclipse.osgi.jar:?]
	at com.liferay.portal.scheduler.internal.SchedulerEngineHelperImpl$SchedulerJobConfigurationServiceTrackerCustomizer.addingService(SchedulerEngineHelperImpl.java:531) ~[?:?]
	at com.liferay.portal.scheduler.internal.SchedulerEngineHelperImpl$SchedulerJobConfigurationServiceTrackerCustomizer.addingService(SchedulerEngineHelperImpl.java:522) ~[?:?]
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:321) ~[org.eclipse.osgi.jar:?]
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:264) ~[org.eclipse.osgi.jar:?]
	at com.liferay.osgi.util.ServiceTrackerFactory.open(ServiceTrackerFactory.java:95) ~[?:?]
	at com.liferay.portal.scheduler.internal.SchedulerEngineHelperImpl.lambda$activate$0(SchedulerEngineHelperImpl.java:347) ~[?:?]
	at com.liferay.portal.kernel.dependency.manager.DependencyManagerSyncUtil.lambda$registerSyncCallable$0(DependencyManagerSyncUtil.java:34) ~[portal-kernel.jar:?]
	at com.liferay.portal.kernel.dependency.manager.DependencyManagerSyncUtil$1.complete(DependencyManagerSyncUtil.java:93) ~[portal-kernel.jar:?]
	at com.liferay.portal.kernel.concurrent.DefaultNoticeableFuture$OnceFutureListener.complete(DefaultNoticeableFuture.java:102) ~[portal-kernel.jar:?]
	at com.liferay.portal.kernel.concurrent.DefaultNoticeableFuture.done(DefaultNoticeableFuture.java:76) ~[portal-kernel.jar:?]
```

Reply:
that's the same thing, the consumer should not be worrying about the null service case. If it happens, it is the supplier's fault. Again, there is no reason to only treat this "BundleContext.getService()" differently. If the null service case is ever going to be considered, hundreds of other usecases will be facing the same thing.

Reply:
@shuyangzhou just to make it clear. Are we OK with the fact that when a scheduled job with an invalid trigger configuration ( interval <= 0) gets deployed the scheduler stops working affecting the rest of the scheduled jobs? We should assume that it's a wrong usage, even if this scheduler jobs are configured through portal property and to fix them it requires to stop the server?

Reply:
It is not the framework core's job to do parameter validation. If you are doing it at the level of property reading which is inside each individual SchedulerJobConfiguration, that's totally fine. But once the execution leaves enter the core, we should trust the incoming service is ready to be consumed.

The issue you show here technically can happen to any ServiceTrackerCustomizer.addingService() that somehow triggers a RuntimeException, and we have 5531 ServiceTracker usage even just from startup, not counting runtime opened ones which would be even more. From a system design point of view, the responsibility must be separated, otherwise we will have a lot of unmaintainable spaghetti code.

Parameter validation should happen at early stage upon receiving the parameters, not at the moment of consuming the services.